### PR TITLE
ctmap: Make GC bpf map dumps more robust.

### PR DESCRIPTION
**Pull request by jrajahalme**
Continue bpf CT map GC iteration from last known found element if the
current element cannot be found. Start again from the beginning
otherwise, but limit restarts from beginning to 2.

Fixes: #3867 
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by ianvernon** [comment]
@jrajahalme I presume you meant to add 'stable/needs-backport' instead of 'stable/backport' so I changed it.

**Comment by ianvernon** [comment]
test-me-please

**Review by borkmann — APPROVED** [review]
Looks fine to me, although my only concern would be that the restart logic still has a potential to fail. Would be nicer to avoid failures altogether, but the subsequent run would then likely catch these elements later. Should this just be a warn instead of error?

Alternative option could be to get rid of ct_delete*() in the BPF prog and handle everything out of golang protecting the (global) ctmap with a (global) mutex on write ops which currently looks not to be the case. Another one could be to collect all unique keys first via getnextkey and drop the entries in a second run based on the collected keys (failed deletes we can ignore then since they were removed already from elsewhere). But I'm okay with either way, potentially need to revisit once we run into the GC restart error often.

**Comment by jrajahalme** [comment]
@borkmann Seems to me that collecting all (unique) keys first still has the same problem with concurrent deletes: The dump will magically restart from the beginning whenever a concurrent delete deletes the next key before we get to call GetNextKey on it? But in this case we would note that we received the same key again (hence "collecting all *unique* keys"). Should we still keep counting how many times we have seen the keys and stop if it takes too long?

Another approach would be to deal with the restarts, but limit the number or total GetNextKey() calls to the configured maximum size for the map. The rationale for this is that we must be ready to iterate a max sized map, i.e., we are already doing that in the worst case, so it should not be so bad if we do that sometimes due to concurrent deletes. With deletes inline the map at least gets smaller as we progress (v.s. collecting keys first)?

**Comment by borkmann** [comment]
Fully agree that with regards to collecting keys, we would still need to deal with restarts, I like the idea of limiting the number of GetNextKey() calls to the max map elems. I was thinking whether it should potentially be higher, but the problem with hashtable is that if the elem was deleted, we really just start over from the beginning, not just from the hashtable slot start, so doesn't make sense to generate more useless map traversal work in that case. So the configured max size sounds about right, and the next scheduled GC cleanup will then hopefully succeed to catch those stale elements.

**Comment by tgraf** [comment]
> Alternative option could be to get rid of ct_delete*() in the BPF prog and handle everything out of golang protecting the (global) ctmap with a (global) mutex on write ops which currently looks not to be the case. Another one could be to collect all unique keys first via getnextkey and drop the entries in a second run based on the collected keys (failed deletes we can ignore then since they were removed already from elsewhere). But I'm okay with either way, potentially need to revisit once we run into the GC restart error often.

The deletion from the BPF prog dramatically improved the behavior on rapid connection open/close cycles. We should keep it.

**Comment by jrajahalme** [comment]
@tgraf Agree. Changed the PR to restart from the beginning if can't continue from the previous element without a restart limit, but limit the total number get next operations to the maximum number of elements in the map and log a warning (instead of an error) if GC cannot be completed.

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/maps/ctmap/ctmap.go [hunk 1]
```diff
@@ -262,44 +282,74 @@ func doGC6(m *bpf.Map, filter *GCFilter) int {
...
+		// remember the last found key
+		prevKey = currentKey
+		prevKeyValid = true
+		// continue from the next key
+		currentKey = nextKey
 	}
+
+	if count == m.MapInfo.MaxEntries {
+		// TODO Add a metric we can bump and observe here.
+		log.WithError(err).Warning("Garbage collection on IPv6 CT map failed to finish")
```
Comment (borkmann) [inline pkg/maps/ctmap/ctmap.go]: Hmm, should that (+ number of iterations) be `count == m.MapInfo.MaxEntries + 1` instead, so we know for sure that there must have been a restart in between? E.g. worst case we could have deleted a full map with all invalid entries, and still trigger this warn, no?

Comment (borkmann) [reply pkg/maps/ctmap/ctmap.go]: Scratch my comment, it's fine as is. LGTM.

Comment (jrajahalme) [reply pkg/maps/ctmap/ctmap.go]: I had `count` being 0-based and while it was correct, it was misleading, fixed.

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/maps/ctmap/ctmap.go (hunk 1 above)
Comment (borkmann) [reply pkg/maps/ctmap/ctmap.go]: Ok, thanks. LGTM

**Comment by eloycoto** [comment]
test-me-please

**Comment by raybejjani** [comment]
backport https://github.com/cilium/cilium/pull/3982

**Pull request merged**